rust/lance-core/src/io/writer/statistics.rs

```diff
@@ -49,75 +55,385 @@ pub struct StatisticsRow {
+
+    let array_iterator = arrays.iter().map(|x| x.as_string::<T>());
+
+    for array in array_iterator {
+        null_count += array.null_count() as i64;
+        if array.null_count() == array.len() {
+            continue;
+        }
+
+        array.iter().for_each(|value| {
+            if let Some(mut val) = value {
+                for i in (BINARY_PREFIX_LENGTH..val.len() + 4).rev() {
+                    if val.is_char_boundary(i) {
+                        val = &val[..i];
+                        break;
+                    }
+                }
+
+                if let Some(v) = min_value {
+                    if let Some(Ordering::Less) = val[..].partial_cmp(v) {
+                        min_value = Some(val);
+                    }
+                } else {
+                    min_value = Some(val);
+                }
+
+                if let Some(v) = max_value {
+                    if let Some(Ordering::Greater) = val.partial_cmp(v) {
+                        max_value = Some(val);
+                    }
+                } else {
+                    max_value = Some(val);
+                }
+            }
+        });
+    }
+
+    if let Some(v) = min_value {
+        if v.len() > BINARY_PREFIX_LENGTH {
+            min_value = truncate_utf8(v, BINARY_PREFIX_LENGTH);
+        }
+    }
+
+    let max_value_bound: Vec<u8>;
+    if let Some(v) = max_value {
+        if v.len() > BINARY_PREFIX_LENGTH {
+            max_value = truncate_utf8(v, BINARY_PREFIX_LENGTH);
+            max_value_bound = increment_utf8(max_value.unwrap().as_bytes().to_vec()).unwrap();
+            max_value = Some(str::from_utf8(&max_value_bound).unwrap());
+        }
+    }
+
+    let min_value = min_value.map(|x| x.to_string());
+    let max_value = max_value.map(|x| x.to_string());
+
+    match arrays[0].data_type() {
+        DataType::Utf8 => StatisticsRow {
+            null_count,
+            min_value: ScalarValue::Utf8(min_value),
+            max_value: ScalarValue::Utf8(max_value),
+        },
+        DataType::LargeUtf8 => StatisticsRow {
+            null_count,
+            min_value: ScalarValue::LargeUtf8(min_value),
+            max_value: ScalarValue::LargeUtf8(max_value),
+        },
+        _ => {
+            todo!()
+        }
+    }
+}
+
+fn get_binary_statistics<T: OffsetSizeTrait>(arrays: &[&ArrayRef]) -> StatisticsRow {
+    let mut min_value: Option<&[u8]> = None;
+    let mut max_value: Option<&[u8]> = None;
+    let mut null_count: i64 = 0;
+
+    let array_iterator = arrays.iter().map(|x| as_generic_binary_array::<T>(x));
+
+    for array in array_iterator {
+        null_count += array.null_count() as i64;
+        if array.null_count() == array.len() {
+            continue;
+        }
+
+        array.iter().for_each(|value| {
+            if let Some(val) = value {
+                // don't compare full buffers if possible
+                let val = &val[..std::cmp::min(BINARY_PREFIX_LENGTH + 4, val.len())];
```

Comment:
Similar thing here. Leave a comment explaining what goes wrong if we don't add 4.

Reply:
Switched to `truncate_binary` and added an explanation, though I'm not happy with the wording:
```
  // Truncate binary buffer to BINARY_PREFIX_LENGTH to avoid comparing potentially
  // very long buffers.
  if val.len() > BINARY_PREFIX_LENGTH {
      // Truncate to longer length to have a valid comparison
      val = truncate_binary(val, BINARY_PREFIX_LENGTH + 4).unwrap();
  }
```

Reply:
> added an explanation, though I'm not happy with the wording

If you test without the +4, what happens? I partially ask this to have you test whether it is really necessary in the first place. If we test and it's not, then we can remove. But if there is something that goes wrong, then you can describe that in the comment.